openjdk/barriers/mmtkObjectBarrier.cpp

```diff
@@ -67,17 +70,21 @@ void MMTkObjectBarrierSetAssembler::object_reference_write_post(MacroAssembler*
   __ andptr(tmp2, 1);
   __ cmpptr(tmp2, 1);
   __ jcc(Assembler::notEqual, done);
+#endif
 
   __ movptr(c_rarg0, obj);
-  __ lea(c_rarg1, dst);
-  __ movptr(c_rarg2, val == noreg ?  (int32_t) NULL_WORD : val);
-  __ call_VM_leaf_base(FN_ADDR(MMTkBarrierSetRuntime::object_reference_write_slow_call), 3);
+  __ xorptr(c_rarg1, c_rarg1);
+  // Note: If `compensate_val_reg == true && UseCompressedOops === true`, the `val` register will be
+  // holding a compressed pointer to the target object. If the write barrier needs to know the
+  // target, we will need to decompress it before passing it to the barrier slow path. However,
+  // since we know the semantics of `ObjectReference`, i.e. it logs the object without looking at
```

Comment:
Did you mean "the semantics of object remembering barrier"?

We need to a check to know this is object remembering barrier somewhere in the code. We can get it from https://github.com/mmtk/mmtk-core/blob/a75309373cd301acad473bbb8ea04670facd6e2a/src/plan/plan_constraints.rs#L35, and store the value somewhere for the C++ code to access.

Reply:
> Did you mean "the semantics of object remembering barrier"?

Oops. That's a typo.  I did mean `ObjectBarrier`.

> We need to a check to know this is object remembering barrier somewhere in the code.

The changed code is in `mmtkObjectBarrier.cpp` which is dedicated to the object remembering barrier.  The class `MMTkObjectBarrierSetAssembler` is only used if `get_selected_barrier()` in `mmtkBarrierSet.cpp` returns `new MMTkObjectBarrier();`  Therefore we don't need to check here, as long as the `get_selected_barrier()` chooses the right barrier.